Contributing.md

```diff
@@ -111,6 +111,33 @@ To include your changes in the release notes, you should create one (or more) ne
 - `removed` - for features which have been removed
 - `fixed` - for "changed" features which were classed as a bugfix
 
+### Style guide
+
+#### Generic code
+
+PyO3 has a lot of generic APIs to increase usability. These can come at the cost of generic code bloat. Where reasonable, try to implement a concrete sub-portion of generic functions. See `get_item` and `get_item_impl` in `dict.rs` as examples of this pattern.
```

Comment:
Across the code I think we've now used `_impl` and `_inner` suffixes, and also just `_` prefix.

I suggest we pick one of the three and I'll push a commit to update the codebase to match. I propose the `_` prefix, because it's short and implies some level of privacy.

Reply:
Hhhmmm, personally, I actually prefer the

```
fn foobar<...>(...) {
  fn inner/impl(...) {
    ...
  }

  inner/impl(..)
}
```

where the privacy is realized by tucking away the non-generic inner function within the outer function.

I am not sure if `_foobar` as the name of the inner function is helpful?

Of course, this style also does not work if the non-generic implementation is used in multiple outer functions.

Reply:
I agree actually, though I'd proposed `_` prefix thinking that combined with outer functions can lead to smallest diffs if the implementation needs to be reused later.

Given that we both prefer `fn inner`, I will adjust this to state preference for that, and suggest the `_` prefix when implementation needs to be shared.